test/spec/Feature/Query/UpdateSpec.hs

```diff
@@ -535,11 +541,137 @@ spec = do
+      baseTable "bulk_update_items" "id" tblDataBeforeBulk
+      `mutatesWith`
+      requestMutation methodPatch "/bulk_update_items?limit=2&offset=1&order=id"
+        [("Prefer", "params=multiple-objects")]
+        [json|[
+          { "name": "item", "observation": "Lost item" }
+        , { "name": "item-2 - 2nd", "observation": "Damaged item" }
+        , { "name": "item-3 - 3rd", "observation": null }
+        ]|]
+      `shouldMutateInto`
+      [json|[
+        { "id": 1, "name": "item-1", "observation": null }
+      , { "id": 2, "name": "item", "observation": "Lost item" }
+      , { "id": 3, "name": "item", "observation": "Lost item" }
+      ]|]
+
+    it "does a full table update without the header for bulk updates, taking only the first item in the json array body" $ do
+      baseTable "bulk_update_items" "id" tblDataBeforeBulk
+      `mutatesWith`
+      requestMutation methodPatch "/bulk_update_items" mempty
+        [json|[
+          { "name": "item", "observation": "Damaged item" }
+        , { "name": "item-3 - 3rd", "observation": null }
+        ]|]
+      `shouldMutateInto`
+      [json|[
+        { "id": 1, "name": "item", "observation": "Damaged item" }
+      , { "id": 2, "name": "item", "observation": "Damaged item" }
+      , { "id": 3, "name": "item", "observation": "Damaged item" }
+      ]|]
+
+    it "rejects a json array that isn't exclusively composed of objects" $
+      request methodPatch "/bulk_update_items"
+          [("Prefer", "tx=commit"), ("Prefer", "params=multiple-objects")]
+          [json|[
+            { "id": 1, "name": "Item 1", "observation": null }
+          , 2
+          , "Item 2"
+          , { "id": 3, "name": "Item 3" }
+          ]|]
+        `shouldRespondWith`
+          [json| {"message":"All object keys must match","code":"PGRST102","hint":null,"details":null} |]
+          { matchStatus  = 400
+          , matchHeaders = [matchContentTypeJson]
+          }
+
+    it "rejects a json array that has objects with different keys" $
+      request methodPatch "/bulk_update_items"
+          [("Prefer", "tx=commit"), ("Prefer", "params=multiple-objects")]
+          [json|[
+            { "id": 1, "name": "Item 1" }
+          , { "id": 2, "name": "Item 2" }
+          , { "id": 3, "name": "Item 3", "observation": "New item" }
+          ]|]
+        `shouldRespondWith`
+          [json| {"message":"All object keys must match","code":"PGRST102","hint":null,"details":null} |]
+          { matchStatus  = 400
+          , matchHeaders = [matchContentTypeJson]
+          }
+
+    it "does not update when no pk is specified in the body" $ do
+      request methodPatch "/bulk_update_items"
+          [("Prefer", "tx=commit"), ("Prefer", "params=multiple-objects")]
+          [json|[
+            { "name": "Item 1" }
+          , { "name": "Item 2" }
+          ]|]
+        `shouldRespondWith` 400
```

Comment:
Here, the error is returned by the database and has this form:

```json
{
    "code": "42703",
    "details": null,
    "hint": null,
    "message": "column pgrst_body.id does not exist"
}
```

It looks a bit confusing, maybe an `SchemaCache` error could be created to specify that this is an issue with bulk updates and no PK is found in the body. @steve-chavez WDYT?

Reply:
@laurenceisla That would be much better :+1: 

Why does it assume an `id` column on the error? Is it hardcoded?

Reply:
> Is it hardcoded?

No, in this case, the PK for the table is `id`. The next test example has a composite PK of `id,name` and only `id` is specified in the body. In that case, the message would be:

```json
{
    "code": "42703",
    "details": null,
    "hint": null,
    "message": "column pgrst_body.name does not exist"
}
```

Reply:
Ah, got it. Q: What does happen when specifying `?columns=id` and the body doesn't have an `id`?

Reply:
It does not give an error unlike normal `PATCH`. That's because the `WHERE` with the primary key is forced in the bulk update:

```sql
WHERE "test"."bulk_update_items"."id" = "pgrst_body"."id"
```

In this case, the `pgrs_body.id` is `null`, so the query works but it won't update any row.

Reply:
I think that's fine. 

> In this case, the pgrs_body.id is null, so the query works but it won't update any row.

I guess one way to prevent this would be to count the array elements and compare it to the updated items. If they're equal then we can assume a unique key was passed for every row, if not, we could throw an error. 

Not sure if worth it though, it could be confusing as well because we can't tell the user which array element would have the incorrect unique.

Reply:
> Not sure if worth it though, it could be confusing as well because we can't tell the user which array element would have the incorrect unique.

Yeah, I'm not sure about this one either, maybe it's better to discuss later for another PR.

> It looks a bit confusing, maybe an `SchemaCache` error could be created to specify that this is an issue with bulk updates and no PK is found in the body.

I mentioned this but, as you said on another chat, using it with `?columns` won't allow to verify if the JSON body has the `id`. So I'll go with the names you suggested maybe `json_body` or `bulk_update_body` instead of `pgrst_body`.